Transfer METS: namespace all elements, include additional metadata

mistydemeo opened the pull request.

Commit:
createMETS: namespace all elements

Commit:
createMETS: add metsHdr, new metadata

* metsHdr includes dashboard UUID information
* Parent METS element gains the transfer UUID

Hwesta commented:
Remove the commented out lines that create Elements, or update those to use namespaces as well?  I don't want someone in the future uncommenting them thinking they work fine.

Other than that, :+1:

Commit:
databaseFunctions: add getAccessionNumberFromTransfer

mistydemeo commented:
Rebased on current qa/1.x, added a new field, and added a new `databaseFunctions` helper to fetch the required info.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS.py:
```diff
@@ -33,6 +33,8 @@
 sys.path.append("/usr/lib/archivematica/archivematicaCommon")
 from archivematicaFunctions import escape
+from databaseFunctions import getAccessionNumberFromTransfer, getUTCDate
+from elasticSearchFunctions import getDashboardUUID
```
I definitely feel like this function is in the wrong place!

Hwesta replied:
It really is.  storageService.py implements a more generic get_setting that is more general than the file it's in.  Move get_setting to archivematicaFunctions and replace getDashboardUUID with a call to that?  Maybe in another PR?

mistydemeo replied:
In another PR, yeah - I'll follow up after this is merged.

Commit:
SIP arrange: fetch file metadata from SS, not ES

The storage service is theoretically a more reliable store of this
information than ES is.

mistydemeo commented:
The primary goal of this was to support fetching file metadata from the SS instead of ES; I've pushed a commit to do this. It depends on artefactual/archivematica-storage-service#59.

This obsoletes the need for #145, since we no longer rely on fetching absolute values from ES.

Commit:
Add new microservice to update transfer METS with a second structMap

refs #7714

mistydemeo commented:
New commit moves the transfer METS generation to after package extraction, as this is the last point at which files may be added or removed to the transfer. Otherwise the storage service won't be informed about certain files.

Commit:
createMETS: only write fileGrp file elements if new

mistydemeo commented:
After discussion with Evelyn and Justin, instead added a second `structMap` that represents the state of the processed transfer. This is the structMap that the SS will consult.

This is ready for review now - tested and working for SIP arrange in several workflows. Tested with standard transfer and DSpace.

Hwesta commented on src/archivematicaCommon/tests/test_database_functions.py:
```diff
@@ -67,3 +67,12 @@ def test_insert_into_event_fetches_correct_agent_from_file(self):
                                            eventIdentifierUUID="event_agent_id")
         assert Event.objects.get(event_id="event_agent_id").linking_agent.name == "Event agent"
+
+    def test_get_accession_number_from_transfer(self):
```
:+1:

Hwesta commented:
:+1:

mistydemeo commented:
I need to clean up history, but I've made a few changes here in order to add `amdSec`s for all events affecting every file in the `processed` `structMap`. This includes name change events. artefactual/archivematica-storage-service#59 now looks up the name change event, and uses the new name if present.

Commit:
createMETS2: move option parser into __main__

With the option parser in the main script body,
it becomes unsafe to import anything from this
script unless the contents of ARGV are exactly as
expected by createMETS2’s parser.

Note that this produces a number of SyntaxWarnings,
and it is still unsafe to use certain functions if
one of these global variables is not defined.
Further cleanup should make these globals into
function parameters.

Commit:
Create amdSecs with events in processed structMap

* Remove unused amdSec creation code from
  createMETS; this existed for future expansion,
  but was never used.
* Create amdSecs in processedStructmap, by using
  the amdSec creation function from createMETS2.
* Extract the iteration code from createMETS into
  a new function, so processedStructmap can
  iterate over transfer contents the same way.

mistydemeo commented:
Rebased and cleaned up history.

Commit:
createMETS2: change globals into function arguments

mistydemeo commented:
Updated the METS2 commit in 61c0b6558993db77d376328f4fdf918a8badc9fe to move the magic globals into function arguments, making it possible to load/import from createMETS2 without a thousand `SyntaxWarning`s.

Commit:
createMETS: move option parsing into __main__

This moves all option parsing into __main__, which makes it safe to
import functions from this file without side effects. Many functions
have been given new arguments to replace the globals they previously
relied on.

Commit:
createMETS: remove unused variable

mistydemeo commented:
Fixed issues in both branches, tested to be working.

Commit:
storageService: add functions to remove, add transfer files

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateProcessedStructmap.py:
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/python -OO
+
+from __future__ import print_function
```
:)

mistydemeo replied:
I'm trying to get in the habit of using it everywhere. :D

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateProcessedStructmap.py:
```diff
@@ -0,0 +1,69 @@
+    return amdSecs
+
+if __name__ == '__main__':
+    from argparse import ArgumentParser
```
Why imported here and not with the rest of the imports?

mistydemeo replied:
It's not used if this isn't the executing script.

Hwesta commented on src/archivematicaCommon/lib/storageService.py:
```diff
@@ -335,3 +342,21 @@ def request_file_deletion(uuid, user_id, user_email, reason_for_deletion):
+    api = _storage_api()
+    api.file(transfer_uuid).contents.delete()
+
+def index_backlogged_transfer_contents(transfer_uuid, file_set):
```
Is this or remove_files_from_transfer used anywhere?

mistydemeo replied:
Yes, they're used by a new script in archivematica-devtools.

mistydemeo replied:
(That's at artefactual/archivematica-devtools#4)

Hwesta replied:
Ah, ok

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateProcessedStructmap.py:
```diff
@@ -0,0 +1,69 @@
+    parser.add_argument("-f", "--fileGroupIdentifier", action="store", dest="fileGroupIdentifier", default="sipUUID")  # transferUUID
+    parser.add_argument("-S", "--sipUUID", action="store", dest="sipUUID", default="")
+    parser.add_argument("-x", "--xmlFile", action="store", dest="xmlFile", default="")
+    (opts, _) = parser.parse_known_args()
```
Is this still necessary, since createMETS and createMETS2 are better able to be imported now?

mistydemeo replied:
Probably not, no! I can change that.

Hwesta commented:
:+1:

Commit:
Update CHANGELOG.md for #7714
